### PR TITLE
Cleanup/Implement Dispose Pattern

### PR DESCRIPTION
**Pull request by rassilon**
See #785 for details, but this fixes the obvious issue in #785 as well as some other problematic looking IDisposable patterns in the other ChromiumWebBrowser flavors.

### CODE REVIEW

**Comment by amaitland** [comment]
> I'm not sure if this is necessary or not in managed C++, but it certainly looked weird for both the Dipose and finalization method to both try and delete _browserSettings.

Should be perfectly safe to delete a `NULL` object. 

https://voidnish.wordpress.com/2006/06/07/you-can-delete-a-null-pointer/

File: CefSharp.OffScreen/ChromiumWebBrowser.cs [hunk 1]
```diff
@@ -58,26 +58,60 @@ public ChromiumWebBrowser(string address, BrowserSettings browserSettings = null
...
+            DownloadHandler = null;
+            KeyboardHandler = null;
+            LifeSpanHandler = null;
+            MenuHandler = null;
+            FocusHandler = null;
+            RequestHandler = null;
+            DragHandler = null;
+            GeolocationHandler = null;
+
+            // Don't reference event listenets any longer:
```
Comment (jankurianski) [inline CefSharp.OffScreen/ChromiumWebBrowser.cs]: listenets -> listeners

**Comment by rassilon** [comment]
But _browsersSettings isn't set to NULL. Thus my paranoia.

**Comment by amaitland** [comment]
> But _browsersSettings isn't set to NULL. Thus my paranoia.

Usually we set unmanaged objects to `NULL` and looking at browser settings closer I see that it's a `struct`, so I'm pretty sure it would be safe to call `delete`. Happy for the extra checking though :+1:

**Comment by amaitland** [comment]
Overall, looks like a great contribution thanks! :+1: 

(The `BrowserSettings` dispose was actually a chance I made just recently, was done in a hurry)

File: CefSharp.Wpf/ChromiumWebBrowser.cs [hunk 2]
```diff
@@ -606,9 +627,6 @@ private string ReadUrlFromDragDropData(IDataObject data, string urlDataFormatNam
             return url.TrimEnd('\0');
         }
 
-
```
Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Stray whitespace that can ideally be removed.

Comment (rassilon) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Are you suggesting I put the white space back? (I did remove that extraneous white space, yes.)

Comment (amaitland) [inline CefSharp.Wpf/ChromiumWebBrowser.cs]: Ignore that, was on the phone, so clearly wasn't paying enough attention! lol

**Comment by amaitland** [comment]
FYI I've updated the title to make it a little more descriptive.

For reference you can include the issue number in one of your commit message and the issue will auto close on merge.
https://help.github.com/articles/closing-issues-via-commit-messages/

**Comment by amaitland** [comment]
I've taken this for a spin and looks solid :+1:

**Pull request merged**